crates/flux-fhir-analysis/src/lib.rs

```diff
@@ -206,10 +226,30 @@ fn assoc_refinement_def(
 
     if let Some(assoc_reft) = assoc_reft {
         let wfckresults = genv.check_wf(impl_id)?;
-        Ok(rty::EarlyBinder(conv::conv_assoc_reft_def(genv, assoc_reft, &wfckresults)?))
+        Ok(Some(rty::EarlyBinder(conv::conv_assoc_reft_def(genv, assoc_reft, &wfckresults)?)))
     } else {
-        Err(QueryErr::InvalidAssocReft { impl_id: impl_id.to_def_id(), name })?
+        Ok(None)
+    }
+}
+
+fn assoc_refinement_def(
+    genv: GlobalEnv,
+    impl_id: LocalDefId,
+    name: Symbol,
+) -> QueryResult<rty::EarlyBinder<rty::Lambda>> {
+    let trait_id = genv.tcx().trait_id_of_impl(impl_id.to_def_id());
+
+    if let Some(lam) = impl_assoc_refinement_def(genv, impl_id, name)? {
+        return Ok(lam);
+    }
+    // TODO: this seems wrong, what if the trait_id is not a local id?
```

Comment:
This bit bothers me... maybe instead one should use some sort of query here?

Reply:
@ranjitjhala, definitely. This should go through the query instead of calling the local function, i.e., we should call `genv.default_assoc_refinement_def(genv, trait_id, name)`. The query will take care of mapping the right id and then redirect it to the `default_assoc_refinement_def` in this file if necessary.